st2actions/st2actions/container/base.py

```diff
@@ -127,6 +128,40 @@ def _do_run(self, runner):
                     (status, result, context) = runner.run(action_params)
                     result = jsonify.try_loads(result)
 
+            LOG.debug('Validating runner output: %s', runner.runner_type.output_schema)
+
+            if runner.runner_type.output_schema.get('unmodeled'):
+                runner.runner_type.output_schema.pop('unmodeled')
+                LOG.warn(
+                    """Deprecation Notice: This runner has previously had unmodeled
+                    output. In StackStorm 3.1 the output will be placed under the
+                    `output` key."""
+                )
+                runner_schema = {
+                    "additionalProperties": True
+                }
+            else:
+                runner_schema = {
+                    "type": "object",
+                    "properties": runner.runner_type.output_schema,
+                    "additionalProperties": False
+                }
+
+            LOG.info("Action Result: %s", result)
+
+            schema.validate(result, runner_schema, cls=schema.get_validator('custom'))
+
+            if runner.action.output_schema:
+                output_schema = copy.deepcopy(runner.runner_type.output_schema)
+                output_schema.update(runner.action.output_schema)
+                action_schema = {
+                    "type": "object",
+                    "properties": output_schema,
+                    "additionalProperties": False
+                }
+                LOG.debug('Validating action output: %s', action_schema)
+                schema.validate(result, action_schema, cls=schema.get_validator('custom'))
```

Comment:
I think it would make sense / make it more user-friendly if we captured original exception and re-throw a more user-friendly one.

E.g. "Failed validation output for action "foo" against local runner schema: blah blah".

Reply:
I didn't see any tests for those scenarios (action output failing runner schema and action output failing action output schema).

Did I miss it, or we don't have any yet? :trollface: